### PR TITLE
Set Application's resource information into GitOpsDeployment's Status field.

### PR DESCRIPTION
**Pull request by jparsai**
# Description
This PR will add a feature to get list of resources created by ArgoCD App and add it into `GitOpsDeployment.Status.Resources` field, to inform users about Health and Synk Status of all resources, which are created for a GitOpsDeployment. 

## Jira Ticket
https://issues.redhat.com/browse/GITOPSRVCE-136
https://issues.redhat.com/browse/GITOPSRVCE-137

## Type of changes

1. Added a new column for `ApplicationState` table in `db-schema.sql`.
2. Added new field for `ApplicationState` type in `db/types`.
3. Updated logic in `application_controller` to get resource list from ArgoCD App and save into table as Byte Array.
4. Added new field for `GitOpsDeploymentStatus` type in `v1alpha1/gitopsdeployment_types`.
5. Updated implementation in `application_event_runner_deployments` to convert resource data from Byte Array to managedgitopsv1alpha1.ResourceStatus object.
6. Added/updated test cases for changes.

### CODE REVIEW

**Review by jgwest — CHANGES_REQUESTED** [review]
Nice, thanks @jparsai! I tried it out and it works great, just a few minor tweaks and good to go.

File: db-schema.sql [hunk 1]
```diff
@@ -262,8 +262,10 @@ CREATE TABLE ApplicationState (
 	-- * Synced
 	-- * OutOfSync
 	-- * Unknown (this is used when Argo CD's status field is "")
-	sync_status VARCHAR (30) NOT NULL
+	sync_status VARCHAR (30) NOT NULL,
 
+	-- resources field comes directly from Argo CD Application CR's .Status.Resources field
+	resources bytea
```
[the remark this answers is not in the document]
Comment (jparsai) [reply db-schema.sql:268]: It seems this is not allowed for bytea. I see this error when I give limit.
```
psql:db-schema.sql:269: ERROR:  type modifier is not allowed for type "bytea"
LINE 25:  resources bytea (262144)
```
I saw few docs and as I understand PgSql doesn't have option to mention limit for bytea.

`PostgreSQL doesn’t limit binary fields. Any migrations adding a :binary field with the :limit option will raise a syntax error.`

Though it says that max limit is 1GB for bytea and with new version it is 2GB.
https://www.postgresql.org/docs/7.4/jdbc-binary-data.html

File: cluster-agent/controllers/argoproj.io/application_controller.go [hunk 2]
```diff
@@ -179,3 +189,33 @@ func (r *ApplicationReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&appv1.Application{}).
 		Complete(r)
 }
+
+// Convert ResourceStatus Array into String and then compress it into Byte Array​
+func CompressResourceData(resources []appv1.ResourceStatus) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (jparsai) [reply cluster-agent/controllers/argoproj.io/application_controller.go:198]: Issue here is that test file is in `argoprojio_test` package and to access function for unit tests I had to export it.
https://github.com/redhat-appstudio/managed-gitops/blob/d68c1af5c55c4adc28e59194b4d7e74b365b805c/cluster-agent/controllers/argoproj.io/application_controller_test.go#L1

File: backend/eventloop/application_event_loop/application_event_runner_deployments.go [hunk 3]
```diff
@@ -13,6 +16,7 @@ import (
 	"github.com/redhat-appstudio/managed-gitops/backend/condition"
 	"github.com/redhat-appstudio/managed-gitops/backend/eventloop/eventlooptypes"
 	"github.com/redhat-appstudio/managed-gitops/backend/util/fauxargocd"
+	"gopkg.in/yaml.v2"
```
[the remark this answers is not in the document]
Comment (jparsai) [reply backend/eventloop/application_event_loop/application_event_runner_deployments.go]: Oh, I didnt notice that it is already there.

File: cluster-agent/controllers/argoproj.io/application_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jgwest) [reply cluster-agent/controllers/argoproj.io/application_controller.go:198]: Gotcha, thanks!

File: db-schema.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (jgwest) [reply db-schema.sql:268]: @jparsai Ah, that's good to know! In that case, let's just add a size check to `CreateApplicationState` and `UpdateApplicationState`, so it returns an error for fields that are larger than that (similar to the `validateFieldLength` function we've already got in there)

File: cluster-agent/controllers/argoproj.io/application_controller.go [hunk 4]
```diff
@@ -114,6 +118,13 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 			applicationState.Revision = db.TruncateVarchar(app.Status.Sync.Revision, db.ApplicationStateRevisionLength)
 			sanitizeHealthAndStatus(applicationState)
 
+			// Get the list of resources created by deployment and convert it into a compressed YAML string.
+			var err error
+			applicationState.Resources, err = CompressResourceData(app.Status.Resources)
+			if err != nil {
+				log.Error(err, "unable to compress resource data into byte array.")
```
Comment (jgwest) [inline cluster-agent/controllers/argoproj.io/application_controller.go:125]: Think you can `return ctrl.Result{}, err` here, unless here is a reason not to.

File: cluster-agent/controllers/argoproj.io/application_controller.go [hunk 5]
```diff
@@ -134,6 +145,9 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	applicationState.Revision = db.TruncateVarchar(app.Status.Sync.Revision, db.ApplicationStateRevisionLength)
 	sanitizeHealthAndStatus(applicationState)
 
+	// Get the list of resources created by deployment and convert it into a compressed YAML string.
+	applicationState.Resources, _ = CompressResourceData(app.Status.Resources)
```
Comment (jgwest) [inline cluster-agent/controllers/argoproj.io/application_controller.go:149]: Can check error here, too, and return on non-nil error.

File: backend/eventloop/application_event_loop/application_event_runner_deployments.go [hunk 6]
```diff
@@ -661,6 +664,13 @@ func (a *applicationEventLoopRunner_Action) applicationEventRunner_handleUpdateD
 	gitopsDeployment.Status.Sync.Status = managedgitopsv1alpha1.SyncStatusCode(applicationState.Sync_Status)
 	gitopsDeployment.Status.Sync.Revision = applicationState.Revision
 
+	// Fetch the list of resources created by deployment from table and update local gitopsDeployment instance.
+	var err error
+	gitopsDeployment.Status.Resources, err = decompressResourceData(applicationState.Resources)
+	if err != nil {
+		a.log.Error(err, "unable to decompress byte array received from table.")
+	}
```
Comment (jgwest) [inline backend/eventloop/application_event_loop/application_event_runner_deployments.go:672]: Should return the error if it is non-nil

**Review by jgwest — CHANGES_REQUESTED** [review]

File: cluster-agent/controllers/argoproj.io/application_controller.go (hunk 4 above)
Comment (jparsai) [reply cluster-agent/controllers/argoproj.io/application_controller.go:125]: Changed it, my thinking here was that we can continue with deployment even though Resource field is not set, just like it is at present.

**Comment by jparsai** [comment]
@jgwest you can review it now, finally it is working. It was not a major change, but today my system was not supporting me :(.

**Review by jgwest — APPROVED** [review]
Great stuff, thanks again @jparsai!

**Pull request merged**